File: java/v1/src/main/java/com/google/displayvideo/api/samples/RetrieveUsers.java
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.displayvideo.api.samples;

import com.beust.jcommander.Parameter;
import com.google.api.client.util.Strings;
import com.google.api.services.displayvideo.v1.DisplayVideo;
import com.google.api.services.displayvideo.v1.DisplayVideo.Users;
import com.google.api.services.displayvideo.v1.DisplayVideoScopes;
import com.google.api.services.displayvideo.v1.model.AssignedUserRole;
import com.google.api.services.displayvideo.v1.model.ListUsersResponse;
import com.google.api.services.displayvideo.v1.model.User;
import com.google.displayvideo.api.samples.utils.ArgumentNames;
import com.google.displayvideo.api.samples.utils.CodeSampleParams;
import java.util.ArrayList;
import java.util.List;

/**
 * This example retrieves accessible Display &amp; Video 360 users based on given filter values.
 *
 * This example makes requests to the Display &amp; Video 360 API Users service that require
 * authentication via service account. Requests made not using a service account will return
 * an error.
 */
public class RetrieveUsers {

  private static class RetrieveUsersParams extends CodeSampleParams {

    @Parameter(
        names = ArgumentNames.EMAIL_ADDRESS,
        description = "String that the email address of retrieved users must contain.")
    public String emailAddress;

    @Parameter(
        names = ArgumentNames.DISPLAY_NAME,
        description = "String that the display name of retrieved users must contain.")
    public String displayName;

    @Parameter(
        names = ArgumentNames.USER_ROLE,
        description = "User role that the retrieved users must have assigned.")
    public String userRole;

    @Parameter(
        names = ArgumentNames.HAS_PARTNER_ROLE,
        description = "Whether the retrieved users must have an assigned user role for a partner.")
    public boolean hasPartnerRole;

    @Parameter(
        names = ArgumentNames.HAS_ADVERTISER_ROLE,
        description =
            "Whether the retrieved users must have an assigned user role for an advertiser.")
    public boolean hasAdvertiserRole;

    @Parameter(
        names = ArgumentNames.PARTNER_ID,
        description = "ID for partner that retrieved users must have a user role assigned for.")
    public String partnerId;

    @Parameter(
        names = ArgumentNames.ADVERTISER_ID,
        description = "ID for advertiser that retrieved users must have a user role assigned for.")
    public String advertiserId;

    @Parameter(
        names = ArgumentNames.PARENT_PARTNER_ID,
        description =
            "ID for parent partner of advertisers that retrieved users must have a user role"
                + " assigned for.")
    public String parentPartnerId;
  }

  public static void main(String[] args) throws Exception {
    RetrieveUsersParams params = new RetrieveUsersParams();

    if (!params.parseArguments(args)) {
      // Either pass the required parameters for this example on the command line, or insert them
      // into the code here. See the parameter class definition above for descriptions.
      params.hasPartnerRole = false;
      params.hasAdvertiserRole = false;
    }

    System.out.println(
        "This function requires the use of a service account and an additional Users service"
            + " scope. These configurations will be applied regardless of specified flags.");

    // Adding user service scope to passed parameter of additional scopes.
    List<String> additionalScopes = new ArrayList<String>();
    if (params.additionalScopes != null && !params.additionalScopes.isEmpty()) {
      additionalScopes = params.additionalScopes;
    }
    additionalScopes.add(DisplayVideoScopes.DISPLAY_VIDEO_USER_MANAGEMENT);

    DisplayVideo service =
        DisplayVideoFactory.getInstance(
            params.clientSecretsFile, true, params.serviceAccountKeyFile, additionalScopes);

    runExample(
        service,
        params.emailAddress,
        params.displayName,
        params.userRole,
        params.hasPartnerRole,
        params.hasAdvertiserRole,
        params.partnerId,
        params.advertiserId,
        params.parentPartnerId);
  }

  public static void runExample(
      DisplayVideo service,
      String emailAddress,
      String displayName,
      String userRole,
      boolean hasPartnerRole,
      boolean hasAdvertiserRole,
      String partnerId,
      String advertiserId,
      String parentPartnerId)
      throws Exception {

    // Create list to store the built filters.
    ArrayList<String> filters = new ArrayList<String>();

    // Build filters with given arguments.
    if (emailAddress != null) {
      filters.add("email:\"" + emailAddress + "\"");
    }
    if (displayName != null) {
      filters.add("displayName:\"" + displayName + "\"");
    }
    if (userRole != null) {
      filters.add("assignedUserRole.userRole=\"" + userRole + "\"");
    }
    if (hasPartnerRole) {
      filters.add("assignedUserRole.entityType=\"PARTNER\"");
    }
    if (hasAdvertiserRole) {
      filters.add("assignedUserRole.entityType=\"ADVERTISER\"");
    }
    if (partnerId != null) {
      filters.add("assignedUserRole.partnerId=\"" + partnerId + "\"");
    }
    if (advertiserId != null) {
      filters.add("assignedUserRole.advertiserId=\"" + advertiserId + "\"");
    }
    if (parentPartnerId != null) {
      filters.add("assignedUserRole.parentPartnerId=\"" + parentPartnerId + "\"");
    }

    // Build full filter string out of filter list.
    String filterStr = String.join(" AND ", filters);

    // Configure the list request.
    Users.List request = service.users().list().setFilter(filterStr);

    // Create the response and nextPageToken variables.
    ListUsersResponse response;
    String nextPageToken = null;

    do {
      // Create and execute the list request.
      response = request.setPageToken(nextPageToken).execute();

      // Check if response is empty.
      if (response.isEmpty()) {
        System.out.print("List request returned no Users");
        break;
      }

      // Iterate over retrieved users.
      for (User user : response.getUsers()) {
        // Print general information about user.
        System.out.printf(
            "User ID: %s, Display name: %s, Email: %s%n",
            user.getUserId(), user.getDisplayName(), user.getEmail());

        // Iterate over and print user's assigned user roles.
        for (AssignedUserRole retrievedRole : user.getAssignedUserRoles()) {
          if (retrievedRole.getPartnerId() != null) {
            System.out.printf(
                "\tPartner ID: %s, Role: %s%n",
                retrievedRole.getPartnerId(), retrievedRole.getUserRole());
          } else if (retrievedRole.getAdvertiserId() != null) {
            System.out.printf(
                "\tAdvertiser ID: %s, Role: %s%n",
                retrievedRole.getAdvertiserId(), retrievedRole.getUserRole());
          }
        }
      }

      // Update the next page token.
      nextPageToken = response.getNextPageToken();
    } while (!Strings.isNullOrEmpty(nextPageToken));
  }
}

File: php/v3/examples/RetrieveUsers.php
<?php
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Require the base class.
require_once __DIR__ . "/../BaseExample.php";

/**
 * This example retrieves accessible Display & Video 360 users based on given
 * filter values.
 *
 * This example makes requests to the Display & Video 360 API Users service that
 * require authentication via service account. Requests made not using a service
 * account will return an error.
 */
class RetrieveUsers extends BaseExample
{
    /**
     * (non-PHPdoc)
     * @see BaseExample::getInputParameters()
     */
    protected function getInputParameters(): array
    {
        return array(
            array(
                'name' => 'email_address',
                'display' => 'Email Address filter value',
                'required' => false
            ),
            array(
                'name' => 'display_name',
                'display' => 'Display Name filter value',
                'required' => false
            ),
            array(
                'name' => 'user_role',
                'display' => 'User Role filter value',
                'required' => false
            ),
            array(
                'name' => 'user_role_entity_type',
                'display' => 'User Role entity type',
                'required' => false,
                'values' => array('PARTNER', 'ADVERTISER')
            ),
            array(
                'name' => 'partner_id',
                'display' => 'Partner Id filter value',
                'required' => false
            ),
            array(
                'name' => 'advertiser_id',
                'display' => 'Advertiser Id filter value',
                'required' => false
            ),
            array(
                'name' => 'parent_partner_id',
                'display' => 'Parent Partner Id filter value',
                'required' => false
            ),
        );
    }

    /**
     * (non-PHPdoc)
     * @see BaseExample::run()
     */
    public function run()
    {
        $values = $this->formValues;
        $emailAddress = $values['email_address'];
        $displayName = $values['display_name'];
        $userRole = $values['user_role'];
        $userRoleEntityType = $values['user_role_entity_type'];
        $partnerId = $values['partner_id'];
        $advertiserId = $values['advertiser_id'];
        $parentPartnerId = $values['parent_partner_id'];

        print '<h2>Users you can access that match filter values</h2>';

        $response = null;
        $nextPageToken = null;

        // Create list to store the built filters.
        $filters = array();

        // Build filters with given arguments.
        if (!empty($emailAddress)) {
            $filters[] = 'email:"' . $emailAddress . '"';
        }
        if (!empty($displayName)) {
            $filters[] = 'displayName:"' . $displayName . '"';
        }
        if (!empty($userRole)) {
            $filters[] = 'assignedUserRole.userRole="' . $userRole .'"';
        }
        if (!empty($userRoleEntityType)) {
            $filters[] =
                'assignedUserRole.entityType="' . $userRoleEntityType . '"';
        }
        if (!empty($partnerId)) {
            $filters[] = 'assignedUserRole.partnerId="' . $partnerId .'"';
        }
        if (!empty($advertiserId)) {
            $filters[] = 'assignedUserRole.advertiserId="' . $advertiserId .'"';
        }
        if (!empty($parentPartnerId)) {
            $filters[] =
                'assignedUserRole.parentPartnerId="' . $parentPartnerId .'"';
        }

        // Build full filter string out of filter list.
        $filterStr = implode(' AND ', $filters);

        do {
            $optParams = array(
                'filter' => $filterStr,
                'pageToken' => $nextPageToken
            );

            // Call the API, retrieving a page of accessible users matching the
            // given filter values.
            try {
                $response = $this
                    ->service
                    ->users
                    ->listUsers(
                        $optParams
                    );
            } catch (\Exception $e) {
                $this->renderError($e);
                return;
            }

            // Print the retrieved users.
            $this->printUsers(
                $response->getUsers()
            );


            // Update the next page token.
            $nextPageToken = $response->getNextPageToken();
        } while (!empty($nextPageToken));
    }

    /**
     * (non-PHPdoc)
     * @see BaseExample::getName()
     */
    public static function getName(): string
    {
        return 'Retrieve Users (Requires service account)';
    }

    /**
     * Prints the given users, their details, and their assigned user roles.
     * @param array $users user resources to be printed.
     */
    protected function printUsers(array $users)
    {
        if (!empty($users)) {
            print '<ul>';

            // Iterate over and print list display of given users.
            foreach ($users as $user) {
                printf(
                    '<li>User %s found:<ul><li>Email address: %s</li><li>'
                        . 'Display name:  %s</li><li>Assigned User Roles:<ul>',
                    $user['name'],
                    $user['email'],
                    $user['displayName']
                );

                // Iterate over and print user's assigned user roles.
                foreach ($user['assignedUserRoles'] as $userRole) {
                    if (isset($userRole['partnerId'])) {
                        printf('<li>Partner ID: %s, User Role: %s</li>',
                            $userRole['partnerId'],
                            $userRole['userRole']
                        );
                    } else if (isset($userRole['advertiserId'])) {
                        printf(
                            '<li>Advertiser ID: %s, User Role: %s</li>',
                            $userRole['advertiserId'],
                            $userRole['userRole']
                        );
                    }
                }
                print '</ul></li></ul></li>';

            }
            print '</ul>';
        } else {
            print '<p>No users returned</p>';
        }
    }
}